### DISCUSSION TITLE
Preserve default total timeout when other timeout is changed in ClientSession.__init__()

### DISCUSSION THREAD

**Discussion by wombatonfire** [aio-libs/aiohttp#7268] [Ideas]
There is a default **total** timeout

`DEFAULT_TIMEOUT: Final[ClientTimeout] = ClientTimeout(total=5 * 60)`

used by the `ClientSession` when no explicit session-level timeout is provided:

```
if timeout is sentinel or timeout is None:
    self._timeout = DEFAULT_TIMEOUT
```

This is, definitely, a good thing, you don't want to wait forever in the worst-case scenario.

The problem is, it's very easy to unintentionally reset this default:

```
timeout = ClientTimeout(connect=5.0)
session = ClientSession(timeout=timeout)
```

Because of the way how ClientSession is initialized

```
if timeout is sentinel or timeout is None:
    self._timeout = DEFAULT_TIMEOUT
else:
    self._timeout = timeout
```

when you change some other - more specific - timeout and do not explicitly provide the value for the total timeout, the default for the total is not preserved and is reset to `None`.

It looks to me like an undesirable side-effect and something you, probably, do not expect. Documentation also does not cover this behavior.

I think the better approach would be to merge the default total timeout value, when it's not being overridden, with whatever more specific timeout is passed on `ClientSession.__init__()`. Something that was done in earlier versions for the now deprecated `read_timeout` and `conn_timeout`:

`self._timeout = attr.evolve(self._timeout, connect=conn_timeout)`

If this makes sense, I'm happy to provide the patch.

**Comment by Dreamsorcerer** [comment]
I think changing the default for `total` in `ClientTimeout` seems like a reasonable approach. Probably will need to be in v4 only though, as this could cause backwards-compatibility issues. Feel free to create a PR for that.

**Reply by wombatonfire** [reply]
@Dreamsorcerer thanks for the feedback! PR #7275 has been created.
